### PR TITLE
leo2, statverif: add darwin support

### PR DESCRIPTION
**Pull request by wegank**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -37162,8 +37162,11 @@ with pkgs;
   elan = callPackage ../applications/science/logic/elan {};
   mathlibtools = with python3Packages; toPythonApplication mathlibtools;
 
-  leo2 = callPackage ../applications/science/logic/leo2
-    { inherit (ocaml-ng.ocamlPackages_4_05) ocaml camlp4; };
+  leo2 = callPackage ../applications/science/logic/leo2 {
+    ocamlPackages = ocaml-ng.mkOcamlPackages (ocaml-ng.ocamlPackages_4_14.ocaml.override {
+      unsafeStringSupport = true;
+    });
+  };
```
Comment (Atemu) [inline pkgs/top-level/all-packages.nix:37169]: Why is this necessary? A comment would be nice.

This smells a tad too complex for all-packages.

Comment (wegank) [reply pkgs/top-level/all-packages.nix:37169]: The reason some packages remain on the obsolete OCaml 4.05 is that this is the last version allowing strings to be mutable. Enabling `unsafeStringSupport` often allows them to bump directly to the last versions in the 4 series, which support aarch64-darwin.

If this is too complex for all-packages, maybe it's time to introduce `ocamlPackages_4_14_unsafe_string`, as [what was done on OPAM](https://ocaml.org/p/ocaml-variants/4.11.2%2Bdefault-unsafe-string)?

Comment (Atemu) [reply pkgs/top-level/all-packages.nix:37169]: I think I'd prefer that. Also with that explanation you just wrote as a comment to explain its existence and need.

Perhaps ocaml 4.05 should probably be aliased to `ocamlPackages_4_14_unsafe_string` on aarch64-darwin.

File: pkgs/top-level/ocaml-packages.nix [hunk 2]
```diff
@@ -1689,6 +1689,10 @@ in let inherit (pkgs) callPackage; in rec
 
   ocamlPackages_4_14 = mkOcamlPackages (callPackage ../development/compilers/ocaml/4.14.nix { });
 
+  ocamlPackages_4_14_unsafe_string = mkOcamlPackages (callPackage ../development/compilers/ocaml/4.14.nix {
+    unsafeStringSupport = true;
+  });
```
Comment (Atemu) [inline pkgs/top-level/ocaml-packages.nix:1694]: Please add a comment why this is useful and/or necessary.

Comment (wegank) [reply pkgs/top-level/ocaml-packages.nix:1694]: I think I'll add comments in `pkgs/development/compilers/ocaml/generic.nix`, and I'm actually wondering if I should do the same for 4.13 (mldonkey) and 4.11 (weidu).

Comment (Atemu) [reply pkgs/top-level/ocaml-packages.nix:1694]: If they're used similarly to the alias you added, by all means! Specifically named attrs like this make the code clearer IMO.

I think I'd prefer a comment next to the attrs themselves though, not some detached place with a similar topic.

**Review by Atemu — APPROVED** [review]
Result of `nixpkgs-review pr 219836` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>leo2</li>
    <li>statverif</li>
  </ul>
</details>
